packages/graphiql-react/src/editor/query-editor.tsx

```diff
@@ -0,0 +1,280 @@
+import type { SchemaReference } from 'codemirror-graphql/utils/SchemaReference';
+import type {
+  FragmentDefinitionNode,
+  GraphQLSchema,
+  ValidationRule,
+} from 'graphql';
+import { MutableRefObject, useContext, useEffect, useRef } from 'react';
+
+import { commonKeys, importCodeMirror } from './common';
+import { EditorContext } from './context';
+import {
+  CompletionCallback,
+  EditCallback,
+  EmptyCallback,
+  useChangeHandler,
+  useCompletion,
+  useKeyMap,
+  useResizeEditor,
+  useSynchronizeValue,
+} from './hooks';
+import { markdown } from '../markdown';
+import { normalizeWhitespace } from './whitespace';
+import { CodeMirrorType, CodeMirrorEditor } from './types';
+
+type OnClickReference = (reference: SchemaReference) => void;
+
+export type UseQueryEditorArgs = {
+  editorTheme?: string;
+  externalFragments?: string | FragmentDefinitionNode[];
+  onClickReference?: OnClickReference;
+  onCopyQuery?: EmptyCallback;
+  onEdit?: EditCallback;
+  onHintInformationRender?: CompletionCallback;
+  onPrettifyQuery?: EmptyCallback;
+  onMergeQuery?: EmptyCallback;
+  onRunQuery?: EmptyCallback;
+  readOnly?: boolean;
+  schema?: GraphQLSchema | null;
+  validationRules?: ValidationRule[];
+  value?: string;
+};
+
+export function useQueryEditor({
+  editorTheme = 'graphiql',
+  externalFragments,
+  onClickReference,
+  onCopyQuery,
+  onEdit,
+  onHintInformationRender,
+  onMergeQuery,
+  onPrettifyQuery,
+  onRunQuery,
+  readOnly = false,
+  schema,
+  validationRules,
+  value,
+}: UseQueryEditorArgs = {}) {
+  const context = useContext(EditorContext);
+  const ref = useRef<HTMLDivElement>(null);
+  const codeMirrorRef = useRef<CodeMirrorType>();
+
+  if (!context) {
+    throw new Error(
+      'Tried to call the `useQueryEditor` hook without the necessary context. Make sure that the `EditorContextProvider` from `@graphiql/react` is rendered higher in the tree.',
+    );
+  }
+
+  const { queryEditor, setQueryEditor } = context;
+
+  const onClickReferenceRef = useRef<OnClickReference>();
+  useEffect(() => {
+    onClickReferenceRef.current = onClickReference;
+  }, [onClickReference]);
+
+  useEffect(() => {
+    importCodeMirror([
+      import('codemirror/addon/comment/comment'),
+      import('codemirror/addon/search/search'),
+      import('codemirror-graphql/esm/hint'),
+      import('codemirror-graphql/esm/lint'),
+      import('codemirror-graphql/esm/info'),
+      import('codemirror-graphql/esm/jump'),
+      import('codemirror-graphql/esm/mode'),
+    ]).then(CodeMirror => {
```

Comment:
Since this effect can be cancelled (deps) do we want to do something like

```
let active = true;
// current stuff
return () => { active = false }
```

This way you can check in your `.then` whether you are in an aborted invocation which could avoid race conditions